web/src/test/java/org/springframework/security/web/bind/support/CurrentSecurityContextArgumentResolverTests.java

```diff
@@ -0,0 +1,239 @@
+
+		User u = (User) resolver.resolveArgument(showSecurityContextWithUserDetail(), null, null,
+				null);
+		assertThat(u.getUsername()).isEqualTo("my_user");
+	}
+
+	@Test
+	public void resolveArgumentSecurityContextErrorOnInvalidTypeImplicit() throws Exception {
+		String principal = "invalid_type_implicit";
+		setAuthenticationPrincipal(principal);
+		assertThat(resolver.resolveArgument(showSecurityContextErrorOnInvalidTypeImplicit(), null, null, null))
+				.isNull();
+	}
+
+	@Test
+	public void resolveArgumentSecurityContextErrorOnInvalidTypeFalse() throws Exception {
+		String principal = "invalid_type_false";
+		setAuthenticationPrincipal(principal);
+		assertThat(resolver.resolveArgument(showSecurityContextErrorOnInvalidTypeFalse(), null, null, null))
+				.isNull();
+	}
+
+	@Test
+	public void resolveArgumentSecurityContextErrorOnInvalidTypeTrue() throws Exception {
+		String principal = "invalid_type_true";
+		setAuthenticationPrincipal(principal);
+		try {
+			resolver.resolveArgument(showSecurityContextErrorOnInvalidTypeTrue(), null, null, null);
+			fail("should not reach here");
+		} catch(ClassCastException ex) {}
+	}
+
+	private MethodParameter showSecurityContextNoAnnotation() {
+		return getMethodParameter("showSecurityContextNoAnnotation", String.class);
+	}
+
+	private MethodParameter showSecurityContextAnnotation() {
+		return getMethodParameter("showSecurityContextAnnotation", SecurityContext.class);
+	}
+
+	private MethodParameter showSecurityContextAuthenticationAnnotation() {
+		return getMethodParameter("showSecurityContextAuthenticationAnnotation", Authentication.class);
+	}
+
+	private MethodParameter showSecurityContextAuthenticationWithOptionalPrincipal() {
+		return getMethodParameter("showSecurityContextAuthenticationWithOptionalPrincipal", Object.class);
+	}
+
+	private MethodParameter showSecurityContextAuthenticationWithPrincipal() {
+		return getMethodParameter("showSecurityContextAuthenticationWithPrincipal", Object.class);
+	}
+
+	private MethodParameter showSecurityContextWithUserDetail() {
+		return getMethodParameter("showSecurityContextWithUserDetail", Object.class);
+	}
+
+	private MethodParameter showSecurityContextErrorOnInvalidTypeImplicit() {
+		return getMethodParameter("showSecurityContextErrorOnInvalidTypeImplicit", String.class);
+	}
+
+	private MethodParameter showSecurityContextErrorOnInvalidTypeFalse() {
+		return getMethodParameter("showSecurityContextErrorOnInvalidTypeFalse", String.class);
+	}
+
+	private MethodParameter showSecurityContextErrorOnInvalidTypeTrue() {
+		return getMethodParameter("showSecurityContextErrorOnInvalidTypeTrue", String.class);
+	}
+
+	private MethodParameter getMethodParameter(String methodName, Class<?>... paramTypes) {
+		Method method = ReflectionUtils.findMethod(TestController.class, methodName,
+				paramTypes);
+		return new MethodParameter(method, 0);
+	}
+
+	public static class TestController {
+		public void showSecurityContextNoAnnotation(String user) {
```

Comment:
Would be nice if one of the methods used a custom implementation for `SecurityContext` so it proves the type matching works.

Reply:
ok, i will add more cases